src/sink/buffer.rs

```diff
@@ -0,0 +1,65 @@
+use std::collections::VecDeque;
+
+use {Poll, Async};
+use sink::{Sink, StartSend, AsyncSink};
+
+/// Sink for the `Sink::buffer` combinator, which buffers up to some fixed
+/// number of values when the underlying sink is unable to accept them.
+#[must_use = "sinks do nothing unless polled"]
+pub struct Buffer<S: Sink> {
+    sink: S,
+    buf: VecDeque<S::SinkItem>,
+    cap: usize,
```

Comment:
I didn't want to rely on the underlying capacity, because you don't get exactly what you asked for, and we don't want the buffer to actually be bigger than what you requested. I'll add a comment.